test/csuite/wt8246_compact_rts_data_correctness/main.c

```diff
@@ -0,0 +1,375 @@
+    large_updates(session, uri, value_b, 30);
+    large_updates(session, uri, value_c, 40);
+    large_updates(session, uri, value_d, 50);
+
+    /* Verify data is visible and correct. */
+    check(session, uri, value_a, 20);
+    check(session, uri, value_b, 30);
+    check(session, uri, value_c, 40);
+    check(session, uri, value_d, 50);
+
+    /* Pin stable to timestamp 40. */
+    testutil_check(__wt_snprintf(tscfg, sizeof(tscfg), "stable_timestamp=%d", 30));
+    testutil_check(conn->set_timestamp(conn, tscfg));
+
+    testutil_check(session->checkpoint(session, NULL));
+
+    /*
+     * Remove 1/3 of data from the middle of the key range to let compact relocate blocks from the
+     * end of the file. Checkpoint the changes after the removal.
+     */
+    remove_records(session, uri, 60);
+
+    /*
+     * Create the compact_started file so that the parent process can start its timer.
+     */
+    sprintf(ckpt_file, compact_file_fmt, home);
+    testutil_checksys((fp = fopen(ckpt_file, "w")) == NULL);
+    testutil_checksys(fclose(fp) != 0);
+
+    run_compact(session, uri);
+}
+static void
+check(WT_SESSION *session, const char *uri, char *value, int read_ts)
+{
+    WT_CURSOR *cursor;
+    size_t val_1_size, val_2_size;
+    uint64_t val1, val2, val3;
+    int i;
+    char *str_val;
+    char tscfg[64];
+
+    printf("Checking value : %s...\n", value);
+
+    testutil_check(__wt_snprintf(tscfg, sizeof(tscfg), "read_timestamp=%d", read_ts));
+    testutil_check(session->begin_transaction(session, tscfg));
+
+    val_1_size = strlen(value);
+    testutil_check(session->open_cursor(session, uri, NULL, NULL, &cursor));
+
+    for (i = 0; i < NUM_RECORDS; i++) {
+        cursor->set_key(cursor, i + 1);
+        testutil_check(cursor->search(cursor));
+        testutil_check(cursor->get_value(cursor, &val1, &val2, &val3, &str_val));
+
+        val_2_size = strlen(str_val);
+
+        testutil_assert(val_1_size == val_2_size);
+        // printf ("str_val : %s\n", str_val);
+        testutil_assert(memcmp(value, str_val, val_1_size) == 0);
+    }
+
+    testutil_check(session->commit_transaction(session, NULL));
+    testutil_check(cursor->close(cursor));
+    cursor = NULL;
+}
+
+static void
+large_updates(WT_SESSION *session, const char *uri, char *value, int commit_ts)
+{
+    WT_CURSOR *cursor;
+    WT_RAND_STATE rnd;
+    uint64_t val;
+    int i;
+    char tscfg[64];
+
+    __wt_random_init_seed((WT_SESSION_IMPL *)session, &rnd);
+    testutil_check(session->open_cursor(session, uri, NULL, NULL, &cursor));
+
+    for (i = 0; i < NUM_RECORDS; i++) {
+        testutil_check(session->begin_transaction(session, NULL));
+        cursor->set_key(cursor, i + 1);
+        val = (uint64_t)__wt_random(&rnd);
+        cursor->set_value(cursor, val, val, val, value);
+        testutil_check(cursor->insert(cursor));
+        testutil_check(__wt_snprintf(tscfg, sizeof(tscfg), "commit_timestamp=%d", commit_ts));
```

Comment:
Same here, it's safe to move the timestamp string formatting outside of the loop.

Reply:
Fixed